tests/tests.py

```diff
@@ -6008,6 +5937,51 @@ def test_plantcv_threshold_texture():
     else:
         assert 0
 
+def create_test_img(sz_img):
+    img = np.random.randint(np.prod(sz_img), size=sz_img) * 255
+    img = img.astype(np.uint8)
+    return img
+
+def create_test_img_bin(sz_img):
+    img = np.zeros(sz_img)
+    img[3:7, 2:8] = 1
+    return img
+@pytest.mark.parametrize("img, refimg, method, refimg_, pts, refpts, debug_mode", [
+    [create_test_img((12,10,3)), create_test_img((12,10,3)), "default", None, [(0,0),(1,0),(0,3),(4,4)], [(0,0),(1,0),(0,3),(4,4)],"plot"],
+    [create_test_img((10,10,3)), create_test_img((11,11)), "lmeds", 1, [(0,0),(1,0),(0,3),(4,4)], [(0,0),(1,0),(0,3),(4,4)],"print"],
+    [create_test_img_bin((10,10)), create_test_img((11,11)), "rho", None, [(0,0),(1,0),(0,3),(4,4)], [(0,0),(1,0),(0,3),(4,4)],"print"],
+    [create_test_img_bin((10,10)), create_test_img((11,11,3)), "ransac", None, [(0,0),(1,0),(0,3),(4,4)], [(0,0),(1,0),(0,3),(4,4)],"print"],
+    [create_test_img((100, 150)), create_test_img((10, 15)), "ransac", None, [(0, 0), (149, 0), (99, 149), (0, 99), (3, 3)], [(0, 0), (0, 14), (9, 14), (0, 9), (3, 3)],"plot"]
+                                                                       ])
+def test_plantcv_transform_warp(img, refimg, method, refimg_, pts , refpts, debug_mode, tmpdir):
+    # Test cache directory
+    # cache_dir = os.path.join(TEST_TMPDIR, "test_plantcv_transform_warp")
+    # os.mkdir(cache_dir)
+    pcv.params.debug_outdir = tmpdir
+    pcv.params.debug = debug_mode
+    warped_img, mat = pcv.transform.warp(img, refimg, pts, refpts, method=method)
+
+    if refimg_ is not None:
+        refimg_ = refimg
+        warped_img_ = pcv.transform.warp_align(img, mat, refimg_)
+        assert warped_img.shape[0:2] == refimg.shape[0:2] and warped_img_.shape[0:2] == refimg_.shape[0:2]
+    else:
+        warped_img_ = pcv.transform.warp_align(img, mat)
+        assert warped_img.shape[0:2] == refimg.shape[0:2] and warped_img_.shape[0:2] == img.shape[0:2]
+
+@pytest.mark.parametrize("img, refimg, pts, refpts, method", [
```

Comment:
- [x ] fail on different length pts
- [x ] fail on same length but <4
- [ ] fail because can't calc homography?

Reply:
maybe we can annotate the different parameterizations for future reference?

Reply:
I agree.
Yes, the 3rd failed because of not being able to converge.